core/shared/src/main/scala/zio/internal/FiberContext.scala

```diff
@@ -963,7 +966,7 @@ private[zio] final class FiberContext[E, A](
       val oldState = state.get
 
       oldState match {
-        case Executing(Status.Suspended(oldStatus, true, _, _, _), observers, interrupted) =>
+        case Executing(Status.Suspended(oldStatus, true, _, _, _), observers, interrupted) if !oldState.interrupting =>
```

Comment:
There is a race condition in `if !oldState.interrupting`. I think it's ok. 

We may want to add a test for this if one doesn't exist, basically that async finalizers during winddown can't be interrupted. Could be tested with a latch.

Reply:
`kill` is called from other fibers, so what happens if the fiber itself changes the state of `setInterrupting` to false at the same time this check is happening. In this case, this code here (in `kill`) will make a different decision about what to do based on the changing boolean value.

Reply:
Right. Then `state` would not be equal to `oldState` and we would loop and then on the next loop `interrupting` would be `false` and we would interrupt. But isn't that the behavior we want? You're definitely right there can be a race between setting the interrupting status and interrupting but I think with this whatever the interrupting status is when the `compareAndSet` gets evaluated will control which seems correct.

Reply:
Added a test for this.

Reply:
Ah, you're right, because you are looking at `oldState.interrupting`, which will be consistent with the state. 

Good with me. 👍